File: Python/sorting/heap_sort.py
from sort_helper import *


def heap_sort(arr, sort):
    __build_heap(arr, sort)
    for i in range(len(arr) - 1, -1, -1):
        arr[0], arr[i] = arr[i], arr[0]
        if sort == 1:
            __max_heapify(arr, i, 0)
        else:
            __min_heapify(arr, i, 0)


def __build_heap(arr, sort):
    bmrmin = len(arr) - 2 // 2  # bottom most right most internal node
    for i in range(bmrmin, -1, -1):
        # For sorting in Ascending Order
        if sort == 1:
            __max_heapify(arr, len(arr), i)
        # For sorting in Descending Order
        else:
            __min_heapify(arr, len(arr), i)


def __max_heapify(arr, n, index):
    l_child, r_child, largest = (
        2 * index + 1,
        2 * index + 2,
        index,
    )
    if l_child < n and arr[l_child] > arr[largest]:
        largest = l_child
    if r_child < n and arr[r_child] > arr[largest]:
        largest = r_child
    if largest != index:
        arr[index], arr[largest] = arr[largest], arr[index]
        __max_heapify(arr, n, largest)


def __min_heapify(arr, n, index):
    l_child, r_child, smallest = (
        2 * index + 1,
        2 * index + 2,
        index,
    )
    if l_child < n and arr[l_child] < arr[index]:
        smallest = l_child
    if r_child < n and arr[r_child] < arr[smallest]:
        smallest = r_child
    if smallest != index:
        arr[index], arr[smallest] = arr[smallest], arr[index]
        __min_heapify(arr, n, smallest)


if __name__ == "__main__":
    arr = take_input()
    heap_sort(arr, sort())
    print_sorted_array(arr)

File: Java/sorting/HeapSort.java
package sorting;

public class HeapSort {
    public static void heapSort(int[] arr, int sort) {
        buildHeap(arr, arr.length, sort);
        for (int i = arr.length - 1; i > 0; i--) {
            swap(arr, 0, i);
            if (sort == 1)
                maxHeapify(arr, i, 0);
            else
                minHeapify(arr, i, 0);
        }
    }

    private static void buildHeap(int[] arr, int n, int sort) {
        int bmrmin = n - 2 / 2; // bottom most right most internal node
        for (int i = bmrmin; i >= 0; i--) {
            // For sorting in Ascending Order
            if (sort == 1)
                maxHeapify(arr, n, i);
            // For sorting in Descending Order
            else
                minHeapify(arr, n, i);
        }
    }

    private static void swap(int[] arr, int a, int b) {
        int temp = arr[b];
        arr[b] = arr[a];
        arr[a] = temp;
    }

    private static void maxHeapify(int[] arr, int n, int index) {
        int largest = index, lChild = 2 * index + 1, rChild = 2 * index + 2;
        if (lChild < n && arr[lChild] > arr[largest])
            largest = lChild;
        if (rChild < n && arr[rChild] > arr[largest])
            largest = rChild;
        if (largest != index) {
            swap(arr, index, largest);
            maxHeapify(arr, n, largest);
        }
    }

    private static void minHeapify(int[] arr, int n, int index) {
        int smallest = index, lChild = 2 * index + 1, rChild = 2 * index + 2;
        if (lChild < n && arr[lChild] < arr[smallest])
            smallest = lChild;
        if (rChild < n && arr[rChild] < arr[smallest])
            smallest = rChild;
        if (smallest != index) {
            swap(arr, index, smallest);
            minHeapify(arr, n, smallest);
        }
    }

    public static void main(String[] args) {
        int[] arr = SortHelper.takeInput();
        heapSort(arr, SortHelper.sort());
        SortHelper.printSortedArray(arr);
    }
}
